bftengine/src/bftengine/ControlStateManager.cpp

```diff
@@ -14,12 +14,32 @@
 #include "ControlStateManager.hpp"
 namespace bftEngine {
 
-void ControlStateManager::setStopAtNextCheckpoint() {}
+// This method will be called from the command handler. Notice that the command handler does not aware to the number of
+// the checkpoint. Thus, it can only raise a flag to stop.
+void ControlStateManager::setStopAtNextCheckpoint() {
+  std::stringstream outStream;
+  controlStateMessages::StopAtNextCheckpointMessage msg(true);
+  concord::serialize::Serializable::serialize(outStream, msg);
+  auto data = outStream.str();
+  state_transfer_.saveReservedPage(
+      resPageOffset() + reserved_pages_indexer_.update_reserved_page_, data.size(), data.c_str());
+}
 
-std::optional<uint64_t> ControlStateManager::getStopCheckpointToStopAt() { return UINT64_MAX; }
+std::optional<bool> ControlStateManager::getStopCheckpointToStopAt() {
+  if (!state_transfer_.loadReservedPage(
+          resPageOffset() + reserved_pages_indexer_.update_reserved_page_, size_of_reserved_pages_, scratchPage_)) {
+    return {};
+  }
+  std::string str(scratchPage_, size_of_reserved_pages_);
+  std::stringstream inStream;
+  inStream.str(str);
+  controlStateMessages::StopAtNextCheckpointMessage msg;
+  concord::serialize::Serializable::deserialize(inStream, msg);
+  return msg.stop_;
+}
 
 ControlStateManager::ControlStateManager(IStateTransfer& state_transfer) : state_transfer_{state_transfer} {
-  state_transfer_.getStatus();  // Temporary, to escape not-used compilation error
+  scratchPage_ = (char*)std::malloc(size_of_reserved_pages_);
```

Comment:
Do we really need manual memory management?
The same effect can be achieved by using a regular `std::string`.
1. It has out-of-the-box integration with `std::stringsream`.
2. We can remove `malloc` and `free`.
3. The dtor can be `=default`.

In the constructor the size of the string can be set like this: `scratchPage_.resize(size_of_reserved_pages_);`
Once the string has to be passed to the `serialize` functions, we may pass `scratchPage_.data()`.
Note, `data()` returns non-constant pointer to char.

Reply:
done